OpenRA.Mods.Common/Traits/Render/WithHarvestAnimation.cs

```diff
@@ -9,64 +9,79 @@
+		[SequenceReference] public readonly string HarvestSequence = "harvest";
 
-		public readonly string Palette = "effect";
-
-		public object Create(ActorInitializer init) { return new WithHarvestAnimation(init.Self, this); }
+		public object Create(ActorInitializer init) { return new WithHarvestAnimation(init, this); }
 	}
 
-	class WithHarvestAnimation : INotifyHarvesterAction
+	public class WithHarvestAnimation : ITick, INotifyHarvesterAction
 	{
-		WithHarvestAnimationInfo info;
-		Animation anim;
-		bool visible;
+		public readonly WithHarvestAnimationInfo Info;
+		readonly WithSpriteBody wsb;
+		readonly Harvester harv;
+
+		public bool IsModifying;
+
+		public WithHarvestAnimation(ActorInitializer init, WithHarvestAnimationInfo info)
+		{
+			Info = info;
+			harv = init.Self.Trait<Harvester>();
+			wsb = init.Self.Trait<WithSpriteBody>();
+		}
 
-		public WithHarvestAnimation(Actor self, WithHarvestAnimationInfo info)
+		protected virtual string NormalizeHarvesterSequence(Actor self, string baseSequence)
 		{
-			this.info = info;
-			var rs = self.Trait<RenderSprites>();
-			var body = self.Trait<IBodyOrientation>();
+			var desiredState = harv.Fullness * (Info.PrefixByFullness.Length - 1) / 100;
+			var desiredPrefix = Info.PrefixByFullness[desiredState];
+
+			if (wsb.DefaultAnimation.HasSequence(desiredPrefix + baseSequence))
+				return desiredPrefix + baseSequence;
+			else
+				return baseSequence;
+		}
 
-			anim = new Animation(self.World, rs.GetImage(self), RenderSprites.MakeFacingFunc(self));
-			anim.IsDecoration = true;
-			anim.Play(info.Sequence);
-			rs.Add(new AnimationWithOffset(anim,
-				() => body.LocalToWorld(info.Offset.Rotate(body.QuantizeOrientation(self, self.Orientation))),
-				() => !visible,
-				() => false,
-				p => ZOffsetFromCenter(self, p, 0)), info.Palette);
+		public void Tick(Actor self)
+		{
+			if (!IsModifying && !string.IsNullOrEmpty(wsb.Info.Sequence) && wsb.DefaultAnimation.HasSequence(NormalizeHarvesterSequence(self, wsb.Info.Sequence)))
+			{
+				if (wsb.DefaultAnimation.CurrentSequence.Name != NormalizeHarvesterSequence(self, wsb.Info.Sequence))
+					wsb.DefaultAnimation.ReplaceAnim(NormalizeHarvesterSequence(self, wsb.Info.Sequence));
+			}
 		}
 
 		public void Harvested(Actor self, ResourceType resource)
 		{
-			if (visible)
-				return;
+			if (!IsModifying && !string.IsNullOrEmpty(Info.HarvestSequence) && wsb.DefaultAnimation.HasSequence(NormalizeHarvesterSequence(self, Info.HarvestSequence)))
+			{
+				IsModifying = true;
+				wsb.PlayCustomAnimation(self, NormalizeHarvesterSequence(self, Info.HarvestSequence), () => IsModifying = false);
+			}
+		}
+
+		// If IsModifying isn't set to true, the docking animation
+		// will be overridden by the WithHarvestAnimation fullness modifier.
+		public void Docked()
+		{
+			IsModifying = true;
+		}
 
-			visible = true;
-			anim.PlayThen(info.Sequence, () => visible = false);
+		public void Undocked()
+		{
+			IsModifying = true;
```

Comment:
This must be the problem. It should be set to false, but only AFTER the undock animation has finished, so simply setting ```IsModifying = false``` here won't do the trick as it breaks/prematurely aborts the undock animation.